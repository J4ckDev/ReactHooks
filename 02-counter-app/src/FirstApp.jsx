import React from "react";
// import React, { Fragment } from "react";
import PropTypes from "prop-types";

const FirstApp = ({ quien, info }) => {
  const saludo = "Hola Mundo";
  const persona = { name: "Jonathan", age: 25 };

  /*
    Usando Fragment   
        return (
            <Fragment>
                <h1>{saludo}</h1>
                <p>Mi primera aplicación</p>
            </Fragment>
        ); 
    */

  // Usando la forma corta sin importar fragment
  return (
    <>
      <h1>{saludo}</h1>
      <p>
        React imprime todo lo que sea texto, números y listas, concatenado su
        contenido; pero en el caso de los booleanos, null y undefined no los
        muestra y en el caso de los objetos toca hacer un procedimiento
        adicional para convertirlo a texto y poderlo mostrar.
      </p>
      <pre>{JSON.stringify(persona, null, 3)}</pre>
      <h2>
        {quien} 
      </h2>
      <p>{info}</p>
    </>
  );
};

// Manejar los tipos de propiedades necesarias para el componente y cuales son necesarias
FirstApp.propTypes = {
  quien: PropTypes.string,
  info: PropTypes.string.isRequired,
};

// Aunque se pueden definir valores por defecto en las variables de la función (const FirstApp = ({ quien = "No hay nadie", info })), 
// otra alternativa es definir las propiedades por defecto del componente, esto nos permite llevar
// un mejor control de las propiedades que necesita el componente

FirstApp.defaultProps = {
    quien : 'No hay nadie',
    info : 'No hay info'
}

export default FirstApp;
