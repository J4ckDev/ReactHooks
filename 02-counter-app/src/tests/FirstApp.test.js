import React from "react";
//import { render } from "@testing-library/react";
import FirstApp from "../FirstApp";
import { shallow } from "enzyme";

describe("Pruebas en el componente <PrimeraApp />", () => {
  /*
    test("Debe mostrar el mensaje Soy J4ckDev", () => {    
        const saludo = "Soy J4ckDev";
        const {getByText} = render(<FirstApp quien="Soy J4ckDev" info=""/>);
        expect(getByText(saludo)).toBeInTheDocument();    
    });
    */
  test("Debe mostrar <FirstApp /> correctamente", () => {
    const wrapper = shallow(<FirstApp />);
    expect(wrapper).toMatchSnapshot();
    // El toMatchSnapshot permite comparar la parte visual de un componenete, de este modo podemos
    // cerciorarnos que si tenemos una alteración en la UI del componente, la prueba fallará y
    // podremos revisarcuales fueron los cambios. Si el cambio es el esperado, se puede generar
    // una nueva instantánea o snapshot del componente. Más info en https://jestjs.io/docs/snapshot-testing
  });
  test("Debe mostrar la info enviada por props", () => {
    const info = "Soy la info";
    const wrapper = shallow(<FirstApp info={info} />);
    const textParagraph = wrapper.find("p").get(1).props.children;
    //wrapper.find() Equivale a document.querySelector()
    //en este caso get sirve para obtener el segundo nodo de parrafo, que contiene la información de info
    expect(textParagraph).toBe(info)
  });
});
//Documentación de enzyme https://enzymejs.github.io/enzyme/
//Adapatador no oficial de enzyme para react 17 https://github.com/wojtekmaj/enzyme-adapter-react-17
