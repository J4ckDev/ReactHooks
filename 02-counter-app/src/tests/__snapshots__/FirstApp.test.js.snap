// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <PrimeraApp /> Debe mostrar <FirstApp /> correctamente 1`] = `
<Fragment>
  <h1>
    Hola Mundo
  </h1>
  <p>
    React imprime todo lo que sea texto, n√∫meros y listas, concatenado su contenido; pero en el caso de los booleanos, null y undefined no los muestra y en el caso de los objetos toca hacer un procedimiento adicional para convertirlo a texto y poderlo mostrar.
  </p>
  <pre>
    {
   "name": "Jonathan",
   "age": 25
}
  </pre>
  <h2>
    No hay nadie
  </h2>
  <p>
    No hay info
  </p>
</Fragment>
`;
